llmfoundry/callbacks/hf_checkpointer.py

```diff
@@ -235,158 +276,175 @@ def _save_checkpoint(self, state: State, logger: Logger):
                 Path(self.save_dir_format_str) /
                 self.huggingface_folder_name_fstr), state.run_name,
             state.timestamp)
-        dir_context_mgr = tempfile.TemporaryDirectory(
-        ) if self.remote_ud is not None else contextlib.nullcontext(
-            enter_result=save_dir)
-
-        with dir_context_mgr as temp_save_dir:
-            assert isinstance(temp_save_dir,
-                              str)  # pyright doesn't know about enter_result
-
-            log.debug('Gathering state dict')
-            from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
-
-            if state.is_model_ddp:
-                composer_model = state.model.module
-                original_model: PreTrainedModel = state.model.module.model
-                state_dict_model = state.model.module.model
-                original_tokenizer = state.model.module.tokenizer
-            elif isinstance(state.model.model, FSDP):
-                composer_model = state.model
-                original_model: PreTrainedModel = state.model.model.module
-                state_dict_model = state.model.model
-                original_tokenizer = state.model.tokenizer
+
```

Comment:
a lot of these code "changes" below are because:
* we now create the temp directory not using a context, so the indentation has changed. This is because we need to keep these files alive for `_register_model_with_run_id_multiprocess`, so now we manually clean up the temp directory at the end of the run along with waiting for the processes to finish. 
* I've added an extra `dist.barrier()` between HF checkpoint upload and mlflow registry logic.